### DISCUSSION TITLE
Doubt about state-> position, velocity and attitude.

### DISCUSSION THREAD

**Discussion by Wallat1** [bitcraze/discussions#297] [Q&A] [answered]
Hi.

I have the following doubts in what part do you assign values to  state->position, state.velocity and state->attitude?, and where do you calculate the state.velocity?, since in my case the positioning system only gives position and attitude.

Regards.

**Comment by krichardsson** [comment]
Hi,

Would it be possible for you to describe your set up a bit more and what you want to achieve? "state" is a pretty generic term that can be interpreted in many different ways, do you mean the state of the estimator?

**Reply by Wallat1** [reply]
Hi.

I mean to these code lines:

148 -struct vec statePos = mkvec(state->position.x, state->position.y, state->position.z);

 149- struct vec stateVel = mkvec(state->velocity.x, state->velocity.y, state->velocity.z);

186- desiredYaw = state->attitude.yaw + setpoint->attitudeRate.yaw * dt;

All of them inside the controller_mellinger.c. On the other hand, in these lines 

262- ew.x = radians(setpoint->attitudeRate.roll) - stateAttitudeRateRoll;

263- ew.y = -radians(setpoint->attitudeRate.pitch) - stateAttitudeRatePitch;

264- ew.z = radians(setpoint->attitudeRate.yaw) - stateAttitudeRateYaw;

setpoint->attitudeRate.roll and setpoint->attitudeRate.pitch take as desired value 0 or where are they calculated?, I ask this since to setpoint->attitudeRate.yaw for mean of the command self._cf.commander.send_setpoint(roll, pitch, yawrate, thrust) can be given a desired value.

Regards.

**Comment by knmcguire** [comment, marked as answer]
What is implemented in the Crazyflie's firmware is a Extendend Kalman Filter. which is able to estimate a state, here defined as global Position.x/y/z, global  Velocity.x/y/z and attitude.roll/pitch/yaw, based on different measurements that does need to be in same axis system (body fixed coordinates versus absolute coordinates) or exactly in the same differential state (flowdeck motion vectors (velocity) -> position, or Mocap (pos) -> velocity). That is all handled by providing a motion (system) model, and a measurement model for each of these measurents how the relate to the main 'state')

You will need to study a bit on what (extended) kalman filters are to understand this completely but wikipedia and youtube can help with that. [This page](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/) explains a bit what state estimators exists within the crazyflie and refers to the papers where they originated from.

ps: attitudeRate is taken almost directly from the gyroscope with minimal filtering.

**Reply by Wallat1** [reply]
Hi.

Thank you so much for your help.

Regards.
